cherrytree c++ uses very new macOS libc++ features...only for 10.15+ at present

I was working on building this with a 10.13 system I had on my lap, but there are some quite new libc++ features being asked for. The Xcode compiler (1000 series) could not handle it, so I installed llvm9, which does have the required include files, etc, but it looks like the system features on 10.13 are too old.  The general portability will be limited, as it is:
```
[ 83%] Building CXX object src/ct/CMakeFiles/cherrytree_shared.dir/ct_actions_edit.cc.o
[ 84%] Building CXX object src/ct/CMakeFiles/cherrytree_shared.dir/ct_actions_export.cc.o
[ 84%] Building CXX object src/ct/CMakeFiles/cherrytree_shared.dir/ct_actions_file.cc.o
[ 84%] Building CXX object src/ct/CMakeFiles/cherrytree_shared.dir/ct_actions_find.cc.o
In file included from /Users/Shared/projects/cherrytree/future/src/ct/ct_actions_export.cc:24:
/Users/Shared/projects/cherrytree/future/src/ct/ct_export2html.h:98:37: error: 'path' is unavailable:
      introduced in macOS 10.15
void to_html(const std::filesystem::path& file, std::ostream& output);
                                    ^
/usr/local/opt/llvm@9/bin/../include/c++/v1/filesystem:738:24: note: 'path' has been explicitly marked
      unavailable here
class _LIBCPP_TYPE_VIS path {
                       ^
[ 85%] Building CXX object src/ct/CMakeFiles/cherrytree_shared.dir/ct_actions_format.cc.o
```

Not sure if there is a way to gear down to some fallback features if these are not available on the system.

--- a/future/src/ct/ct_actions_import.cc
+++ b/future/src/ct/ct_actions_import.cc
@@ -28,9 +28,6 @@
 #include "ct_logging.h"
 #include <fstream>
 
-namespace fs = std::filesystem;
-
-
 // Import a node from a html file
 void CtActions::import_node_from_html_file() noexcept 
 {

--- a/future/src/ct/ct_export2html.cc
+++ b/future/src/ct/ct_export2html.cc
@@ -31,7 +31,6 @@
 #include <fstream>
 #include <filesystem>
 
-namespace fs = std::filesystem;
 
 CtExport2Html::CtExport2Html(CtMainWin* pCtMainWin)
  : _pCtMainWin(pCtMainWin)
@@ -694,14 +693,14 @@ void to_html(std::istream& input, std::ostream& output, std::string from_format)
 }
 
 
-void to_html(const std::filesystem::path& file, std::ostream& output) {
+void to_html(const std::string& file, std::ostream& output) {
     auto process = pandoc_process();
-    process->append_arg(file.string());
+    process->append_arg(file);
 
     try {
         process->run(output);
     } catch(std::exception& e) {
-        spdlog::error("Exception in to_html with filepath: {}; message: {}", file.string(), e.what());
+        spdlog::error("Exception in to_html with filepath: {}; message: {}", file, e.what());
         throw;
     }
 }

--- a/future/src/ct/ct_export2html.h
+++ b/future/src/ct/ct_export2html.h
@@ -105,6 +105,6 @@ constexpr bool supports_syntax(std::string_view syntax) {
 
 void to_html(std::istream& input, std::ostream& output, std::string from_format);
 
-void to_html(const std::filesystem::path& file, std::ostream& output);
+void to_html(const std::string& file, std::ostream& output);
 
 } // CtPandoc

--- a/future/src/ct/ct_imports.cc
+++ b/future/src/ct/ct_imports.cc
@@ -28,8 +28,6 @@
 #include <fstream>
 #include <sstream>
 
-namespace fs = std::filesystem;
-
 namespace CtXML {
 
 xmlpp::Element* codebox_to_xml(xmlpp::Element* parent, const Glib::ustring& justification, int char_offset, int frame_width, int frame_height, int width_in_pixels, const Glib::ustring& syntax_highlighting, bool highlight_brackets, bool show_line_numbers) 
@@ -171,14 +169,14 @@ std::vector<std::pair<int, int>> CtImports::get_web_links_offsets_from_plain_tex
 std::unique_ptr<ct_imported_node> CtImports::traverse_dir(const std::string& dir, CtImporterInterface* importer)
 {
     auto dir_node = std::make_unique<ct_imported_node>(dir, Glib::path_get_basename(dir));
-    for (const auto& dir_item: fs::directory_iterator(dir))
+    for (const auto& dir_item: CtFileSystem::get_dir_entries(dir))
     {
-        if (fs::is_directory(dir_item))
+        if (Glib::file_test(dir_item, Glib::FILE_TEST_IS_DIR))
         {
-            if (auto node = traverse_dir(dir_item.path().string(), importer))
+            if (auto node = traverse_dir(dir_item, importer))
               dir_node->children.emplace_back(std::move(node));
         }
-        else if (auto node = importer->import_file(dir_item.path().string()))
+        else if (auto node = importer->import_file(dir_item))
             dir_node->children.emplace_back(std::move(node));
     }
 
@@ -807,8 +805,7 @@ CtHtmlImport::CtHtmlImport(CtConfig* config) : _config(config)
 
 std::unique_ptr<ct_imported_node> CtHtmlImport::import_file(const std::string& file)
 {
-    std::filesystem::path filepath(file);
-    if (filepath.extension() != ".html" && std::filesystem::path(filepath).extension() != ".htm")
+    if (!str::endswith(file, ".html") && !str::endswith(file, ".htm"))
         return nullptr;
 
     std::ifstream infile;
@@ -817,7 +814,7 @@ std::unique_ptr<ct_imported_node> CtHtmlImport::import_file(const std::string& f
     std::ostringstream ss;
     ss << infile.rdbuf();
 
-    auto imported_node = std::make_unique<ct_imported_node>(file, filepath.stem().string());
+    auto imported_node = std::make_unique<ct_imported_node>(file, CtFileSystem::get_file_stem(file));
     CtHtml2Xml html2xml(_config);
     html2xml.set_local_dir(Glib::path_get_dirname(file));
     html2xml.set_outter_xml_doc(&imported_node->xml_content);
@@ -978,13 +975,12 @@ CtZimImport::CtZimImport(CtConfig* config): CtParser(config), CtTextParser(confi
 
 std::unique_ptr<ct_imported_node> CtZimImport::import_file(const std::string& file)
 {
-    std::filesystem::path filepath(file);
-    if (filepath.extension() != ".txt")
+    if (!str::endswith(file, ".txt"))
         return nullptr;
 
-    _ensure_notebook_file_in_dir(filepath.parent_path());
+    _ensure_notebook_file_in_dir(Glib::path_get_dirname(file));
 
-    std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, filepath.stem().string());
+    std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, CtFileSystem::get_file_stem(file));
     _current_node = node.get();
     _current_element = node->xml_content.create_root_node("root")->add_child("slot");
     _current_element = _current_element->add_child("rich_text");
@@ -1148,18 +1144,18 @@ void CtZimImport::_parse_body_line(const std::string& line)
     _add_newline();
 }
 
-void CtZimImport::_ensure_notebook_file_in_dir(const fs::path& dir)
+void CtZimImport::_ensure_notebook_file_in_dir(const std::string& dir)
 {
     if (_has_notebook_file) return;
-    for (auto dir_item: fs::directory_iterator(dir))
-        if (dir_item.path().filename() == "notebook.zim")
+    for (auto dir_item: CtFileSystem::get_dir_entries(dir))
+        if (Glib::path_get_basename(dir_item) == "notebook.zim")
         {
             _has_notebook_file = true;
             break;
         }
 
     if (!_has_notebook_file) {
-        throw CtImportException(fmt::format("Directory: {} does not contain a notebook.zim file", dir.string()));
+        throw CtImportException(fmt::format("Directory: {} does not contain a notebook.zim file", dir));
     }
 }
 
@@ -1178,7 +1174,7 @@ std::unique_ptr<ct_imported_node> CtPlainTextImport::import_file(const std::stri
         std::ostringstream data;
         data << infile.rdbuf();
 
-        std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, fs::path(file).stem().string());
+        std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, CtFileSystem::get_file_stem(file));
         node->xml_content.create_root_node("root")->add_child("slot")->add_child("rich_text")->add_child_text(data.str());
         node->node_syntax = CtConst::PLAIN_TEXT_ID;
         return node;
@@ -1199,7 +1195,7 @@ CtMDImport::CtMDImport(CtConfig* config) : _parser(config)
 
 std::unique_ptr<ct_imported_node> CtMDImport::import_file(const std::string& file)
 {
-    if (fs::path(file).extension() != ".md")
+    if (!str::endswith(file, ".md"))
         return nullptr;
 
     std::ifstream infile(file);
@@ -1207,7 +1203,7 @@ std::unique_ptr<ct_imported_node> CtMDImport::import_file(const std::string& fil
     _parser.wipe();
     _parser.feed(infile);
 
-    std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, fs::path(file).stem().string());
+    std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, CtFileSystem::get_file_stem(file));
     node->xml_content.create_root_node_by_import(_parser.get_root_node());
 
     return node;
@@ -1225,7 +1221,7 @@ std::unique_ptr<ct_imported_node> CtPandocImport::import_file(const std::string&
     std::stringstream html_buff;
     CtPandoc::to_html(file, html_buff);
 
-    std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, fs::path(file).stem().string());
+    std::unique_ptr<ct_imported_node> node = std::make_unique<ct_imported_node>(file, CtFileSystem::get_file_stem(file));
 
     CtHtml2Xml parser(_config);
     parser.set_outter_xml_doc(&node->xml_content);

--- a/future/src/ct/ct_imports.h
+++ b/future/src/ct/ct_imports.h
@@ -235,7 +235,7 @@ protected:
 
 private:
     void _parse_body_line(const std::string& line);
-    void _ensure_notebook_file_in_dir(const std::filesystem::path& dir);
+    void _ensure_notebook_file_in_dir(const std::string& dir);
 
 private:
     bool              _has_notebook_file {false};

--- a/future/src/ct/ct_misc_utils.cc
+++ b/future/src/ct/ct_misc_utils.cc
@@ -319,7 +319,7 @@ URI_TYPE get_uri_type(const std::string &uri) {
     constexpr std::array<std::string_view, 2> http_ids = {"https://", "http://"};
     constexpr std::array<std::string_view, 2> fs_ids = {"/", "C:\\\\"};
     if (str::startswith_any(uri, http_ids)) return URI_TYPE::WEB_URL;
-    else if (str::startswith_any(uri, fs_ids) || std::filesystem::exists(uri)) return URI_TYPE::LOCAL_FILEPATH;
+    else if (str::startswith_any(uri, fs_ids) || Glib::file_test(uri, Glib::FILE_TEST_EXISTS)) return URI_TYPE::LOCAL_FILEPATH;
     else return URI_TYPE::UNKNOWN;
 }
 }
@@ -971,6 +971,25 @@ int CtFileSystem::getsize(const std::string& path)
     if (g_stat(path.c_str(), &st) == 0)
         return st.st_size;
     return 0;
+}
+
+std::list<std::string> CtFileSystem::get_dir_entries(const std::string& dir)
+{
+    Glib::Dir gdir(dir);
+    std::list<std::string> entries(gdir.begin(), gdir.end());
+    for (auto& entry: entries)
+        entry = Glib::build_filename(dir, entry);
+    return entries;
+}
+
+std::string CtFileSystem::get_file_stem(const std::string& path)
+{
+    if (path == "") return "";
+    std::string name = Glib::path_get_basename(path);
+    size_t dot_pos = name.find_last_of(".");
+    if (dot_pos == std::string::npos || dot_pos == 0)
+        return name;
+    return name.substr(0, dot_pos);
 }
 
 // Open Filepath with External App

--- a/future/src/ct/ct_misc_utils.h
+++ b/future/src/ct/ct_misc_utils.h
@@ -422,6 +422,9 @@ time_t getmtime(const std::string& path);
 
 int getsize(const std::string& path);
 
+std::list<std::string> get_dir_entries(const std::string& dir);
+std::string get_file_stem(const std::string& path);
+
 void external_filepath_open(const std::string& filepath, bool open_fold_if_no_app_error);
 void external_folderpath_open(const std::string& folderpath);
 

--- a/future/src/ct/ct_p7za_iface.cc
+++ b/future/src/ct/ct_p7za_iface.cc
@@ -76,6 +76,7 @@ int CtP7zaIface::p7za_extract(const gchar* input_path, const gchar* out_dir, con
                 "-bd",   // Disable progress indicator
                 "-bso0", // Disable standard output, error output is turn on
                 "-bsp0", // Disable progress output
+                "-bse1", // redirect error output into standard output
                 "-y",
                 "-o" + std::string(out_dir),
                 input_path

--- a/future/tests/tests_misc_utils.cpp
+++ b/future/tests/tests_misc_utils.cpp
@@ -343,6 +343,25 @@ TEST(MiscUtilsGroup, mime__type_contains)
 #endif
 }
 
+
+
+TEST_GROUP(FileSystemGroup)
+{
+};
+
+
+TEST(FileSystemGroup, get_file_stem)
+{
+#ifndef _WIN32
+    STRCMP_EQUAL("", CtFileSystem::get_file_stem("").c_str());
+    STRCMP_EQUAL("file", CtFileSystem::get_file_stem("/root/file").c_str());
+    STRCMP_EQUAL("file", CtFileSystem::get_file_stem("/root/file.txt").c_str());
+    STRCMP_EQUAL("file.2", CtFileSystem::get_file_stem("/root/file.2.txt").c_str());
+    STRCMP_EQUAL(".txt", CtFileSystem::get_file_stem("/root/.txt").c_str());
+#endif
+}
+
+
 int main(int ac, char** av)
 {
     // libp7za has memory leaks

